okio/src/commonMain/kotlin/okio/FileHandle.kt

```diff
@@ -0,0 +1,219 @@
+  /** Pushes all buffered bytes to their final destination. */
+  @Throws(IOException::class)
+  abstract fun flush()
+
+  /**
+   * Returns a source that reads from this starting at [offset]. The returned source must be closed
+   * when it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun source(offset: Long = 0L): Source {
+    synchronized(this) {
+      check(!closed) { "closed" }
+      openStreamCount++
+    }
+    return FileHandleSource(this, offset)
+  }
+
+  /**
+   * Returns the position of [source] in the file. The argument [source] must be either a source
+   * produced by this file handle, or a [BufferedSource] that directly wraps such a source. If the
+   * parameter is a [BufferedSource], it adjusts for buffered bytes.
+   */
+  @Throws(IOException::class)
+  fun position(source: Source): Long {
+    var source = source
+    var bufferSize = 0L
+
+    if (source is RealBufferedSource) {
+      bufferSize = source.buffer.size
+      source = source.source
+    }
+
+    require(source is FileHandleSource && source.fileHandle === this) {
+      "source was not created by this FileHandle"
+    }
+
+    return source.position - bufferSize
+  }
+
+  /**
+   * Returns a sink that writes to this starting at [offset]. The returned sink must be closed when
+   * it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun sink(offset: Long = 0L): Sink {
+    synchronized(this) {
+      check(!closed) { "closed" }
+      openStreamCount++
+    }
+    return FileHandleSink(this, offset)
+  }
+
+  /**
+   * Returns a sink that writes to this starting at the end. The returned sink must be closed when
+   * it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun appendingSink(): Sink {
+    return sink(size())
+  }
+
+  /**
+   * Returns the position of [sink] in the file. The argument [sink] must be either a sink produced
+   * by this file handle, or a [BufferedSink] that directly wraps such a sink. If the parameter is a
+   * [BufferedSink], it adjusts for buffered bytes.
+   */
+  @Throws(IOException::class)
+  fun position(sink: Sink): Long {
+    var sink = sink
+    var bufferSize = 0L
+
+    if (sink is RealBufferedSink) {
+      bufferSize = sink.buffer.size
+      sink = sink.sink
+    }
+
+    require(sink is FileHandleSink && sink.fileHandle === this) {
+      "sink was not created by this FileHandle"
+    }
+
+    return sink.position + bufferSize
+  }
+
+  @Throws(IOException::class)
+  override fun close() {
+    synchronized(this) {
+      if (closed) return@close
+      closed = true
+      if (openStreamCount != 0) return@close
+    }
+    closeInternal()
+  }
+
+  /**
+   * Subclasses should implement this to release resources held by this file handle. It is invoked
+   * once both the file handle is closed, and also all sources and sinks produced by it are also
+   * closed.
+   */
+  @Throws(IOException::class)
+  protected abstract fun closeInternal()
+
+  private class FileHandleSink(
```

Comment:
"inner class" skip the val constructor param?

Reply:
I will, if you can help me figure out how to make this work in the enclosing class:

```
    require(source is FileHandleSource && source.fileHandle === this) {
      "source was not created by this FileHandle"
    }
```

Reply:
It could work if you did something like this I think.
```kotlin
  inner class FileHandleSink(
    var position: Long
  ) : Sink {
    var fileHandle = this@FileHandle
```

But I don't care either way how you end up writing it. I like constructor param.

Reply:
[https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS40LjMwIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsImpzQ29kZSI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiY2xhc3MgQSB7XG4gICAgZnVuIGIoKSA9IEIoKVxuICAgIFxuICAgIGlubmVyIGNsYXNzIEIge1xuICAgICAgICBmdW4gcnVuKCkge1xuICAgICAgICAgICAgcHJpbnRsbih0aGlzKVxuICAgICAgICAgICAgcHJpbnRsbih0aGlzQEEpXG4gICAgICAgICAgICBwcmludGxuKHRoaXNAQilcbiAgICAgICAgfVxuICAgIH1cbn1cblxuZnVuIG1haW4oKSB7XG4gICAgQSgpLmIoKS5ydW4oKVxufSJ9](play.kotlinlang.org)

```
class A {
    fun b() = B()
    
    inner class B {
        fun run() {
            println(this)
            println(this@A)
            println(this@B)
        }
    }
}

fun main() {
    A().b().run()
}
```

Reply:
How about this one?
https://pl.kotl.in/GiG6aSoGI

Reply:
Yeah, your solution seems cleaner.